src/index/CompressedBlockPrefiltering.h

```diff
@@ -0,0 +1,140 @@
+
+#include "global/Id.h"
+#include "global/ValueIdComparators.h"
+#include "index/CompressedRelation.h"
+
+// The compressed block metadata (see `CompressedRelation.h`) that we use to
+// filter out the non-relevant blocks by checking their content of
+// `firstTriple_` and `lastTriple_` (`PermutedTriple`)
+using BlockMetadata = CompressedBlockMetadata;
+
+namespace prefilterExpressions {
+//______________________________________________________________________________
+/*
+`PrefilterExpression` represents a base class for the following sub-classes that
+implement the block-filtering procedure for the specific relational + logical
+operations/expressions.
+
+Remark: We do not actually evaluate the respective SPARQL Expression. We only
+pre-filter w.r.t. blocks that contain relevant data for the actual evaluation of
+those expressions to make the evaluation procedure more efficient.
+
+The block-filtering is applied with the following operations:
+Relational Expressions - `<=`, `>=`, `<`, `>`, `==` and `!=`
+Logical Operations - `and` and `or`
+*/
+
+class PrefilterExpression {
+ public:
+  // The respective metadata to the blocks is expected to be provided in
+  // a sorted order (w.r.t. the relevant column).
+  virtual std::vector<BlockMetadata> evaluate(std::vector<BlockMetadata>& input,
+                                              size_t evaluationColumn) = 0;
+  virtual ~PrefilterExpression() = default;
+};
+
+//______________________________________________________________________________
+// For the actual comparison of the relevant ValueIds from the metadata triples,
+// we use the implementations from ValueIdComparators.
+//
+// Supported comparisons are:
+//  - LessThan, LessEqual, Equal, NotEqual, GreaterEqual, GreaterThan
+using CompOp = valueIdComparators::Comparison;
+
+//______________________________________________________________________________
+// Given a PermutedTriple, retrieve the suitable Id w.r.t. a column (index).
+constexpr auto getIdFromColumnIndex =
+    [](const BlockMetadata::PermutedTriple& permutedTriple,
+       const size_t columnIndex) {
+      switch (columnIndex) {
+        case 0:
+          return permutedTriple.col0Id_;
+        case 1:
+          return permutedTriple.col1Id_;
+        case 2:
+          return permutedTriple.col2Id_;
+        default:
+          // Triple!
+          AD_FAIL();
+      }
+    };
+
+//______________________________________________________________________________
+template <CompOp Comparison>
+class RelationalExpressions : public PrefilterExpression {
+ public:
+  explicit RelationalExpressions(ValueId referenceId)
+      : referenceId_(referenceId) {}
+
+  std::vector<BlockMetadata> evaluate(std::vector<BlockMetadata>& input,
+                                      size_t evaluationColumn) override;
+
+ private:
+  // The ValueId the comparison refers to.
+  ValueId referenceId_;
+  // Helper function that implements the relational comparison on the block
```

Comment:
Order in a class:
1. private data members
2. public constructors and functions.
3. private member functions.

(otherwise the data members are hidden somewhere in the middle of a class).